File: frontend/src/stores/aiThreads.ts
import {
  createThread,
  deleteThread,
  getThread,
  listThreads,
  sendMessage,
  updateThread,
} from '@/api/aiThreads'
import type { AIMessage, AIThread, AIThreadListItem } from '@/types'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useAIThreadsStore = defineStore('aiThreads', () => {
  const threads = ref<AIThreadListItem[]>([])
  const currentThread = ref<AIThread | null>(null)
  const loading = ref(false)
  const sending = ref(false)

  async function loadThreads(bookId: number) {
    loading.value = true
    try {
      threads.value = await listThreads(bookId)
    } finally {
      loading.value = false
    }
  }

  async function openThread(threadId: number) {
    loading.value = true
    try {
      currentThread.value = await getThread(threadId)
    } finally {
      loading.value = false
    }
  }

  async function startNewThread(bookId: number, title?: string) {
    const thread = await createThread(bookId, title)
    currentThread.value = thread
    // Refresh list
    await loadThreads(bookId)
    return thread
  }

  async function renameThread(threadId: number, title: string) {
    await updateThread(threadId, title)
    if (currentThread.value?.id === threadId) {
      currentThread.value.title = title
    }
    const listItem = threads.value.find((t) => t.id === threadId)
    if (listItem) listItem.title = title
  }

  async function removeThread(threadId: number, bookId: number) {
    await deleteThread(threadId)
    if (currentThread.value?.id === threadId) {
      currentThread.value = null
    }
    threads.value = threads.value.filter((t) => t.id !== threadId)
  }

  async function send(
    content: string,
    contextSectionId?: number | null,
    selectedText?: string | null,
  ): Promise<AIMessage | null> {
    if (!currentThread.value) return null
    sending.value = true
    try {
      // Optimistically add user message
      const userMsg: AIMessage = {
        id: -Date.now(),
        thread_id: currentThread.value.id,
        role: 'user',
        content,
        created_at: new Date().toISOString(),
      }
      currentThread.value.messages.push(userMsg)

      const response = await sendMessage(
        currentThread.value.id,
        content,
        contextSectionId,
        selectedText,
      )

      // Replace optimistic user message with real one and add assistant response
      // The server stores both — re-fetch for accurate state
      currentThread.value = await getThread(currentThread.value.id)
      return response
    } finally {
      sending.value = false
    }
  }

  return {
    threads,
    currentThread,
    loading,
    sending,
    loadThreads,
    openThread,
    startNewThread,
    renameThread,
    removeThread,
    send,
  }
})

File: backend/app/api/routes/ai_threads.py
"""AI Threads API endpoints — CRUD and message sending."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.api.deps import get_db, get_settings
from app.api.schemas import (
    AIMessageCreateRequest,
    AIMessageResponse,
    AIThreadCreateRequest,
    AIThreadListItem,
    AIThreadResponse,
    AIThreadUpdateRequest,
)
from app.config import Settings
from app.db.models import AIMessage, AIThread, Book

router = APIRouter(tags=["ai-threads"])


@router.get("/api/v1/books/{book_id}/ai-threads", response_model=list[AIThreadListItem])
async def list_threads(
    book_id: int,
    db: AsyncSession = Depends(get_db),
):
    """List AI threads for a book, ordered by updated_at DESC."""
    # Verify book exists
    book_result = await db.execute(select(Book).where(Book.id == book_id))
    if not book_result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Book not found")

    result = await db.execute(
        select(AIThread)
        .options(selectinload(AIThread.messages))
        .where(AIThread.book_id == book_id)
        .order_by(AIThread.updated_at.desc())
    )
    threads = list(result.scalars().all())

    return [
        AIThreadListItem(
            id=t.id,
            book_id=t.book_id,
            title=t.title,
            message_count=len(t.messages),
            last_message_preview=t.messages[-1].content[:100] if t.messages else None,
            created_at=t.created_at,
            updated_at=t.updated_at,
        )
        for t in threads
    ]


@router.post(
    "/api/v1/books/{book_id}/ai-threads",
    response_model=AIThreadResponse,
    status_code=201,
)
async def create_thread(
    book_id: int,
    body: AIThreadCreateRequest,
    db: AsyncSession = Depends(get_db),
):
    """Create a new AI thread for a book."""
    book_result = await db.execute(select(Book).where(Book.id == book_id))
    if not book_result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Book not found")

    thread = AIThread(
        book_id=book_id,
        title=body.title,
    )
    db.add(thread)
    await db.commit()
    await db.refresh(thread)

    return AIThreadResponse(
        id=thread.id,
        book_id=thread.book_id,
        title=thread.title,
        messages=[],
        created_at=thread.created_at,
        updated_at=thread.updated_at,
    )


@router.get("/api/v1/ai-threads/{thread_id}", response_model=AIThreadResponse)
async def get_thread(
    thread_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get a thread with all messages."""
    result = await db.execute(
        select(AIThread)
        .options(selectinload(AIThread.messages))
        .where(AIThread.id == thread_id)
    )
    thread = result.scalar_one_or_none()
    if not thread:
        raise HTTPException(status_code=404, detail="Thread not found")
    return thread


@router.patch("/api/v1/ai-threads/{thread_id}", response_model=AIThreadResponse)
async def update_thread(
    thread_id: int,
    body: AIThreadUpdateRequest,
    db: AsyncSession = Depends(get_db),
):
    """Update a thread's title."""
    result = await db.execute(
        select(AIThread)
        .options(selectinload(AIThread.messages))
        .where(AIThread.id == thread_id)
    )
    thread = result.scalar_one_or_none()
    if not thread:
        raise HTTPException(status_code=404, detail="Thread not found")

    thread.title = body.title
    await db.commit()
    await db.refresh(thread)

    # Re-fetch with messages
    result = await db.execute(
        select(AIThread)
        .options(selectinload(AIThread.messages))
        .where(AIThread.id == thread_id)
    )
    thread = result.scalar_one_or_none()
    return thread


@router.delete("/api/v1/ai-threads/{thread_id}", status_code=204)
async def delete_thread(
    thread_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Delete a thread and all its messages."""
    result = await db.execute(
        select(AIThread).where(AIThread.id == thread_id)
    )
    thread = result.scalar_one_or_none()
    if not thread:
        raise HTTPException(status_code=404, detail="Thread not found")

    await db.delete(thread)
    await db.commit()


@router.post(
    "/api/v1/ai-threads/{thread_id}/messages",
    response_model=AIMessageResponse,
    status_code=201,
)
async def send_message(
    thread_id: int,
    body: AIMessageCreateRequest,
    db: AsyncSession = Depends(get_db),
    settings: Settings = Depends(get_settings),
):
    """Send a message in a thread — invokes LLM and returns assistant response."""
    from app.services.ai_thread_service import AIThreadService
    from app.services.summarizer.claude_cli import ClaudeCodeCLIProvider

    llm = ClaudeCodeCLIProvider(
        cli_command=settings.llm.cli_command,
        default_model=settings.llm.model,
        default_timeout=settings.llm.timeout_seconds,
        max_budget_usd=settings.llm.max_budget_usd,
    )
    service = AIThreadService(session=db, llm_provider=llm, settings=settings)

    try:
        assistant_msg = await service.send_message(
            thread_id=thread_id,
            content=body.content,
            context_section_id=body.context_section_id,
            selected_text=body.selected_text,
        )
    except ValueError as err:
        raise HTTPException(status_code=404, detail=str(err)) from err

    return assistant_msg


@router.get(
    "/api/v1/ai-threads/{thread_id}/messages",
    response_model=list[AIMessageResponse],
)
async def list_messages(
    thread_id: int,
    db: AsyncSession = Depends(get_db),
):
    """List messages in a thread."""
    result = await db.execute(
        select(AIThread).where(AIThread.id == thread_id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Thread not found")

    result = await db.execute(
        select(AIMessage)
        .where(AIMessage.thread_id == thread_id)
        .order_by(AIMessage.created_at.asc())
    )
    return list(result.scalars().all())
